lib/MusicBrainz/Server/Controller/Role/Load.pm

```diff
@@ -107,6 +107,9 @@ role
         }
 
         if ($entity) {
+            if ($id_is_guid && $entity->gid ne $id) {
+                $c->res->redirect($c->uri_for($c->action, [$entity->gid]), 301);
```

Comment:
Ah, another thing: you need to include the old query parameters in the new URL (this is especially important for the web service).

```suggestion
                $c->res->redirect($c->uri_for($c->action, [$entity->gid], $c->req->params), 301);
```